src/xrc/src/api.rs

```diff
@@ -300,6 +303,25 @@ fn get_normalized_timestamp(
     }
 }
 
+/// This function returns the given [QueriedExchangeRate] option if it contains at least
+/// [MIN_NUM_RATES_FOR_PRIVILEGED_CANISTERS] rates.
+/// Otherwise, the function returns `None`.
+fn get_rate_for_privileged_canisters(
+    maybe_rate: Option<QueriedExchangeRate>,
+) -> Option<QueriedExchangeRate> {
+    if let Some(ref rate) = maybe_rate {
+        if rate.base_asset.symbol == USDT
+            || rate.rates.len() >= MIN_NUM_RATES_FOR_PRIVILEGED_CANISTERS
+        {
+            Some(rate.to_owned())
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
```

Comment:
How about something like this instead?

```rust
fn get_rate_from_cache(
    cache: &mut ExchangeRateCache,
    caller: &Principal,
    symbol: &str,
    timestamp: u64,
) -> Option<QueriedExchangeRate> {
    let maybe_rate = cache.get(symbol, timestamp);
    match maybe_rate {
        Some(ref rate) => {
            if !utils::is_caller_privileged(caller) {
                return maybe_rate;
            }

            if rate.base_asset.symbol == USDT {
                return maybe_rate;
            }

            if rate.rates.len() >= MIN_NUM_RATES_FOR_PRIVILEGED_CANISTERS {
                Some(rate.to_owned())
            } else {
                None
            }
        }
        None => None,
    }
}
```

Then inside the `with_cache_mut` block you would then just have:

```rust
        let maybe_base_rate =
            get_rate_from_cache(cache, &caller, &request.base_asset.symbol, timestamp.value);
```